packages/hardhat-core/src/types/network.ts

```diff
@@ -0,0 +1,70 @@
+import { EventEmitter } from "events";
+
+import { ResolvedNetworkConfig } from "./config";
+
+export interface RequestArguments {
+  readonly method: string;
+  readonly params?: readonly unknown[] | object;
+}
+
+export interface ProviderRpcError extends Error {
+  code: number;
+  data?: unknown;
+}
+
+export interface ProviderMessage {
+  readonly type: string;
+  readonly data: unknown;
+}
+
+export interface EthSubscription extends ProviderMessage {
+  readonly type: "eth_subscription";
+  readonly data: {
+    readonly subscription: string;
+    readonly result: unknown;
+  };
+}
+
+export interface ProviderConnectInfo {
+  readonly chainId: string;
+}
+
+// TODO-HH: Improve the types
+export interface EIP1193Provider extends EventEmitter {
+  request(args: RequestArguments): Promise<unknown>;
+}
+
+export interface JsonRpcRequest {
+  jsonrpc: string;
+  method: string;
+  params: any[];
+  id: number;
+}
+
+export interface JsonRpcResponse {
+  jsonrpc: string;
+  id: number;
+  result?: any;
+  error?: {
+    code: number;
+    message: string;
+    data?: any;
+  };
+}
+
+export interface EthereumProvider extends EIP1193Provider {
+  send(method: string, params?: any[]): Promise<any>;
+  sendAsync(
+    payload: JsonRpcRequest,
+    callback: (error: any, response: JsonRpcResponse) => void
+  ): void;
+}
+
+// This alias is here for backwards compatibility
+export type IEthereumProvider = EthereumProvider;
+
+export interface Network {
```

Comment:
I think this type belongs to the runtime. I'm not sure if we need to name it actually, or just define it in place.

Reply:
Removing this type altogether would cause places like [this one](https://github.com/nomiclabs/buidler/blob/1d55f66e9bb3f06e15c492c0b25226365ccc31bc/packages/hardhat-core/src/internal/core/runtime-environment.ts#L38) or [this one](https://github.com/nomiclabs/buidler/blob/1d55f66e9bb3f06e15c492c0b25226365ccc31bc/packages/hardhat-waffle/src/waffle-provider-adapter.ts#L9) to be rewritten like this:
```typescript
network: HardhatRuntimeEnvironment["network"]
```
It's not terrible but if someone else were to be using this type it might end up being a bit inconvenient.

Reply:
Good point. Let's keep it then. Just move it to runtime